[iOS] Copying and pasting unnamed PDF data into Mail compose should insert an attachment

whsieh opened the pull request:
#### fe7553ab3123ce571cdb1ee13372711954b8e741
<pre>
[iOS] Copying and pasting unnamed PDF data into Mail compose should insert an attachment
<a href="https://bugs.webkit.org/show_bug.cgi?id=246894">https://bugs.webkit.org/show_bug.cgi?id=246894</a>
rdar://96853551

Reviewed by Megan Gardner and Tim Horton.

For attachment element WebKit clients on iOS, when PDF data is written directly to the pasteboard
without either:

1. `-[NSItemProvider preferredPresentationStyle]` being set to Attachment.
2. A non-null `-[NSItemProvider suggestedName]`.

...we&apos;ll fail to read the PDF data as an attachment element upon pasting. However, copying and
pasting this same data into Notes successfully inserts a PDF attachment. To make our paste handling
consistent with Notes when attachments are enabled, we make all composite content (e.g.
presentations such as Keynote files and Powerpoints, EPUB files, and importantly PDFs) that isn&apos;t
one of the rich text types that we would otherwise insert as inline content (i.e. web archive data
and flattened or unflattened RTF formats) insert themselves as attachment elements.

Test: WKAttachmentTests.PasteRawUnnamedPDFData

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::shouldTreatAsAttachmentByDefault):
(WebCore::prefersAttachmentRepresentation):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:

Add an API test to exercise this change, by writing a PDF data blob directly to the pasteboard
without going through item providers or setting a file name.

Canonical link: <a href="https://commits.webkit.org/255932@main">https://commits.webkit.org/255932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fca0cffc327e3afa1f8a9eb2a3d1ee1c28626713

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94036 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3230 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24587 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103664 "Updated wpe dependencies (failure)") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164015 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98028 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3240 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31460 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86352 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99715 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99696 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2328 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80459 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29344 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84254 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72297 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37848 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17771 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35721 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19035 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4095 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39596 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41533 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38267 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -180,6 +180,16 @@ WK_SYSTEM_CONFIGURATION_LDFLAGS_macosx = -framework SystemConfiguration;
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_iphoneos = $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS$(WK_IOS_14));
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_iphonesimulator = $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS$(WK_IOS_14));
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_maccatalyst = $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS$(WK_IOS_14));
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_IOS_SINCE_14 = -framework UniformTypeIdentifiers;
```
Is this necessary? I certainly don't think trunk builds for iOS 13 anymore.

whsieh replied:
Fixed!

whsieh committed:
[iOS] Copying and pasting unnamed PDF data into Mail compose should insert an attachment
https://bugs.webkit.org/show_bug.cgi?id=246894
rdar://96853551

Reviewed by Megan Gardner and Tim Horton.

For attachment element WebKit clients on iOS, when PDF data is written directly to the pasteboard
without either:

1. `-[NSItemProvider preferredPresentationStyle]` being set to Attachment.
2. A non-null `-[NSItemProvider suggestedName]`.

...we'll fail to read the PDF data as an attachment element upon pasting. However, copying and
pasting this same data into Notes successfully inserts a PDF attachment. To make our paste handling
consistent with Notes when attachments are enabled, we make all composite content (e.g.
presentations such as Keynote files and Powerpoints, EPUB files, and importantly PDFs) that isn't
one of the rich text types that we would otherwise insert as inline content (i.e. web archive data
and flattened or unflattened RTF formats) insert themselves as attachment elements.

Test: WKAttachmentTests.PasteRawUnnamedPDFData

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::shouldTreatAsAttachmentByDefault):
(WebCore::prefersAttachmentRepresentation):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:

Add an API test to exercise this change, by writing a PDF data blob directly to the pasteboard
without going through item providers or setting a file name.

Canonical link: https://commits.webkit.org/255932@main
